compiler/vm/vmjit.nim

```diff
@@ -114,49 +123,129 @@ func removeLastEof(c: var TCtx) =
+  for i, it in peek(data.procedures):
+    c.symToIndexTbl[it.id] = LinkIndex(i)
+
+  for i, it in peek(data.constants):
+    c.symToIndexTbl[it.id] = LinkIndex(i)
+
+  # first register globals, then threadvars. This order must be the same as
+  # the one they're later committed to in
+  for i, it in peek(data.globals):
+    c.symToIndexTbl[it.id] = LinkIndex(i)
+
+  for i, it in peek(data.threadvars):
+    c.symToIndexTbl[it.id] = LinkIndex(i)
+
+proc generateMirCode(c: var TCtx, n: PNode;
+                     isStmt = false): (MirTree, SourceMap) =
+  ## Generates the initial MIR code for a standalone statement/expression.
+  if isStmt:
+    # we want statements wrapped in a scope, hence generating a proper
+    # fragment
+    result = generateCode(c.graph, c.module, selectOptions(c), n)
+  else:
+    generateCode(c.graph, selectOptions(c), n, result[0], result[1])
+
+proc generateAST(c: var TCtx, tree: sink MirTree,
+                 source: sink SourceMap): PNode {.inline.} =
+  if tree.len > 0: generateAST(c.graph, c.idgen, c.module, tree, source)
+  else:            newNode(nkEmpty)
+
+proc genStmt*(jit: var JitState, c: var TCtx; n: PNode): VmGenResult =
+  ## Generates and emits code for the standalone top-level statement `n`.
   c.removeLastEof()
-  c.setupLinkState()
 
-  let n = canonicalizeSingle(c.graph, c.idgen, c.module, n, selectOptions(c))
-  gatherDependencies(c, n)
+  # `n` is expected to have been put through ``transf`` already
+  var (tree, sourceMap) = generateMirCode(c, n, isStmt = true)
+  discoverGlobalsAndRewrite(jit.discovery, tree, sourceMap)
+  discoverFrom(jit.discovery, MagicsToKeep, tree)
+  register(c, jit.discovery)
 
   let
+    n = generateAST(c, tree, sourceMap)
     start = c.code.len
     r = vmgen.genStmt(c, n)
 
   if unlikely(r.isErr):
+    rewind(jit.discovery)
     return VmGenResult.err(r.takeErr)
 
   c.gABC(n, opcEof)
-  updateEnvironment(c)
+  updateEnvironment(c, jit.discovery)
 
   result = VmGenResult.ok: (start: start, regCount: c.prc.regInfo.len)
 
-proc genExpr*(c: var TCtx; n: PNode): VmGenResult =
+proc genExpr*(jit: var JitState, c: var TCtx, n: PNode): VmGenResult =
   ## Generates and emits code for the standalone expression `n`
   c.removeLastEof()
-  c.setupLinkState()
 
-  let n = canonicalizeSingle(c.graph, c.idgen, c.module, n, selectOptions(c))
-  gatherDependencies(c, n)
+  # XXX: the way standalone expressions are currently handled is going to
+  #      be a problem as soon as proper MIR passes need to be run (which
+  #      all expect statements). Ideally, dedicated support for
+  #      expressions would be removed from the JIT.
+
+  var (tree, sourceMap) = generateMirCode(c, n)
+  # constant expression outside of routines can currently also contain
+  # definitions of globals...
+  # XXX: they really should not, but that's up to sem. Example:
+  #
+  #        const c = block: (var x = 0; x)
+  #
+  #     If `c` is defined at the top-level, then `x` is a "global" variable
```

Comment:
Ouch, that's busted. I guess not from a scope perspective but a point of declaration perspective (`owner`)?